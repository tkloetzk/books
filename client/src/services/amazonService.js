import apiConfig from '../config/apiConfig';
import axios from 'axios';

export function getAmazonBookService(isbns) {
  return axios
    .post(apiConfig.amazon, { isbns })
    .then(books => books.data)
    .catch(error => {
      throw error;
    });
}

export function getAmazonBookServiceSingle(isbn) {
  return axios
    .post(apiConfig.amazonV2, { isbn })
    .then(res => res.data)
    .catch(error => {
      throw error;
    });
}
