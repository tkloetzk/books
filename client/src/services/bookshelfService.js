import apiConfig from '../config/apiConfig';
import axios from 'axios';
import sortBooklist from '../util/calculator';

export function getBookshelfService(excludedGenres) {
  return axios
    .post(apiConfig.bookshelf, excludedGenres)
    .then(res => sortBooklist(res.data))
    .catch(error => {
      throw error;
    });
}
export function addBookshelfService(booklist) {
  return axios
    .post(`${apiConfig.bookshelf}/add`, booklist)
    .then(res => console.log(res.data))
    .catch(error => {
      throw error;
    });
}

export function updateBookOnBookshelfService(id, fields) {
  return axios
    .put(`${apiConfig.bookshelf}/update/${id}`, fields)
    .then(res => console.log(res.data))
    .catch(error => {
      throw error;
    });
}
