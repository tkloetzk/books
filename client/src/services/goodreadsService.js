import apiConfig from '../config/apiConfig';
import axios from 'axios';

export function getGoodreadsSingleBooksService(booklist) {
  return axios
    .get(`${apiConfig.goodreads}/${booklist}`)
    .then(booklist => booklist.data)
    .catch(error => {
      throw error;
    });
}
