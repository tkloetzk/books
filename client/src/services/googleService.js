import apiConfig from '../config/apiConfig';
import axios from 'axios';

export function getGoogleBookService(isbn) {
  return axios
    .get(`${apiConfig.google}/${isbn}`)
    .then(book => book.data)
    .catch(error => {
      throw error;
    });
}
