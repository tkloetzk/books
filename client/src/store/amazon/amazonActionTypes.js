export const FETCH_AMAZON_BOOK_HAS_ERRORED = 'FETCH_AMAZON_BOOK_HAS_ERRORED';
export const FETCH_AMAZON_BOOK_IS_LOADING = 'FETCH_AMAZON_BOOK_IS_LOADING';
export const FETCH_AMAZON_BOOK_SUCCESS = 'FETCH_AMAZON_BOOK_SUCCESS';
export const FETCH_SINGLE_AMAZON_BOOK_SUCCESS =
  'FETCH_SINGLE_AMAZON_BOOK_SUCCESS';
export const CLEAR_AMAZON_BOOKS_SUCCESS = 'CLEAR_AMAZON_BOOKS_SUCCESS';
