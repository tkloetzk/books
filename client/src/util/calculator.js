import get from 'lodash/get';
import forEach from 'lodash/forEach'
import clone from 'lodash/clone'

let meanGoodreadsVotes
let meanAmazonVotes
let totalMean

export default function sortBooklist(originalBooklist) {
  const booklist = clone(originalBooklist);
  calculateMeans(booklist);
  calculateAdjustedRating(booklist);
  booklist.sort((a, b) =>
    a.adjustedRating < b.adjustedRating
      ? 1
      : b.adjustedRating < a.adjustedRating
      ? -1
      : 0
  );
  return booklist;
}

function calculateMeans(booklist) {
  let goodreadsVotes = 0,
    amazonVotes = 0,
    total = 0;

  forEach(booklist, book => {
    goodreadsVotes += get(book, 'goodreadsAverageRating')
    amazonVotes += get(book, 'amazonAverageRating');
    total +=
      get(book, 'goodreadsAverageRating') +
      get(book, 'amazonAverageRating');
  })

  meanGoodreadsVotes = goodreadsVotes / booklist.length;
  meanAmazonVotes = amazonVotes / booklist.length;
  totalMean = total / 2 / booklist.length;

  return booklist;
}

function calculateAdjustedRating(booklist) {
  for (var i = 0; i < booklist.length; i++) {
    var book = booklist[i];
    var adjustedRating =
      (getAdjustedRating(
        get(book, 'goodreadsRatingsCount'),
        get(book, 'goodreadsAverageRating'),
        meanGoodreadsVotes,
        trimmean(getGoodreadsRatingsCountList(booklist))
      ) +
        getAdjustedRating(
          get(book, 'amazonRatingsCount'),
          get(book, 'amazonAverageRating'),
          meanAmazonVotes,
          trimmean(getAmazonRatingsCountList(booklist))
        )) /
      2;
    book.adjustedRating = adjustedRating;
  }
}

function getAdjustedRating(ratingsCount, averageRating, meanVote, minVotes) {
  //double rating = averageRating * 2;
  return (
    (ratingsCount / (ratingsCount + minVotes)) * averageRating +
    (minVotes / (ratingsCount + minVotes)) * meanVote
  );
}

function getGoodreadsRatingsCountList(booklist) {
  let votes = [];
  for (const book of booklist) {
    votes.push(book.goodreadsRatingsCount);
  }

  return votes.sort(function(a, b) {
    return a - b;
  });
}

function getAmazonRatingsCountList(booklist) {
  let votes = [];
  for (const book of booklist) {
    votes.push(book.amazonRatingsCount);
  }

  return votes.sort(function(a, b) {
    return a - b;
  });
}

function trimmean(votes) {
  votes.sort(function(a, b) {
    return a - b;
  });

  const removeAmount = (votes.length * 0.25) / 2;
  const roundDownNearedMultipleTwo = Math.floor(
    removeAmount >= 0
      ? (removeAmount / 2) * 2
      : ((removeAmount - 2 + 1) / 2) * 2
  );
  return getMean(
    votes.slice(
      roundDownNearedMultipleTwo,
      votes.length - roundDownNearedMultipleTwo
    )
  );
}

function getMean(array) {
  let sum = 0;
  for (let i = 0; i < array.length; i++) {
    sum += array[i];
  }
  return sum / array.length;
}
