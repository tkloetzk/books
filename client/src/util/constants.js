export const LOADING_STATUSES = {
  initial: 'initial',
  loading: 'loading',
  success: 'success',
  errored: 'errored',
};
