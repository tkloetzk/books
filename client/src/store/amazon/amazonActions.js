import * as types from './amazonActionTypes';
import {
  getAmazonBookService,
  getAmazonBookServiceSingle,
} from '../../services/amazonService';

export function getAmazonBookFailure(bool) {
  return {
    type: types.FETCH_AMAZON_BOOK_HAS_ERRORED,
    hasErrored: bool,
  };
}

export function getAmazonBookIsLoading(bool) {
  return {
    type: types.FETCH_AMAZON_BOOK_IS_LOADING,
    isLoading: bool,
  };
}

export function getAmazonBookSuccess(books) {
  return {
    type: types.FETCH_AMAZON_BOOK_SUCCESS,
    books,
  };
}

export function getSingleAmazonBookSuccess(book) {
  return {
    type: types.FETCH_SINGLE_AMAZON_BOOK_SUCCESS,
    book,
  };
}

export function getAmazonBook(isbns) {
  return dispatch => {
    dispatch(getAmazonBookIsLoading(true));
    return getAmazonBookService(isbns)
      .then(resp => {
        dispatch(getAmazonBookIsLoading(false));
        dispatch(getAmazonBookSuccess(resp.books));
        return resp.books;
      })
      .catch(error => {
        dispatch(getAmazonBookIsLoading(false));
        dispatch(getAmazonBookFailure(true));
      });
  };
}

export function getAmazonSingleBook(isbn) {
  return dispatch => {
    dispatch(getAmazonBookIsLoading(true));
    return getAmazonBookServiceSingle(isbn)
      .then(resp => {
        dispatch(getAmazonBookIsLoading(false));
        dispatch(getSingleAmazonBookSuccess(resp.book));
        return resp.book;
      })
      .catch(error => {
        dispatch(getAmazonBookIsLoading(false));
        dispatch(getAmazonBookFailure(true));
      });
  };
}
