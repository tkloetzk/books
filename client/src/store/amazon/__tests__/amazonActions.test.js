import * as actions from '../amazonActions';
import * as types from '../amazonActionTypes';
import { LOADING_STATUSES } from '../../../util/constants';
import { getAmazonBookServiceSingle } from '../../../services/amazonService';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

const createMockStore = configureMockStore([thunk]);

jest.mock('../../../services/amazonService');

describe('amazon actions', () => {
  let book;
  beforeEach(() => {
    book = {
      amazonAverageRating: 4.4,
      amazonRatingsCount: 929,
      price: '',
      isbn: '9780547564654',
    };
    getAmazonBookServiceSingle.mockClear();
  });
  it('can set amazon book hasErrored', () => {
    const actionTrue = actions.getAmazonBookFailure(true);
    expect(actionTrue).toEqual({
      type: types.FETCH_AMAZON_BOOK_HAS_ERRORED,
      hasErrored: true,
    });

    const actionFalse = actions.getAmazonBookFailure(false);
    expect(actionFalse).toEqual({
      type: types.FETCH_AMAZON_BOOK_HAS_ERRORED,
      hasErrored: false,
    });
  });
  it('can set amazon book isLoading', () => {
    const action = actions.getAmazonBookIsLoading(LOADING_STATUSES.initial);
    expect(action).toEqual({
      type: types.FETCH_AMAZON_BOOK_IS_LOADING,
      isLoading: LOADING_STATUSES.initial,
    });
  });
  it('can set amazon book success', () => {
    const action = actions.getAmazonBookSuccess(book);
    expect(action).toEqual({
      type: types.FETCH_AMAZON_BOOK_SUCCESS,
      books: book,
    });
  });
  it('can clear amazon books', () => {
    const dispatch = jest.fn();
    actions.clearAmazonBooks()(dispatch);
    expect(dispatch.mock.calls).toMatchSnapshot();
  });
  it('can clear amazon books success', () => {
    const action = actions.clearAmazonBooksSuccess();
    expect(action).toEqual({
      type: types.CLEAR_AMAZON_BOOKS_SUCCESS,
    });
  });
  it('can get amazon book', () => {
    const store = createMockStore();
    const response = {
      book: {
        amazonAverageRating: 4.4,
        amazonRatingsCount: 929,
        price: '',
        isbn: '9780547564654',
      },
    };
    const expectedResponse = [
      { isLoading: 'loading', type: 'FETCH_AMAZON_BOOK_IS_LOADING' },
      { isLoading: 'success', type: 'FETCH_AMAZON_BOOK_IS_LOADING' },
      {
        book: {
          amazonAverageRating: 4.4,
          amazonRatingsCount: 929,
          isbn: '9780547564654',
          price: '',
        },
        type: 'FETCH_AMAZON_BOOK_SUCCESS',
      },
    ];
    getAmazonBookServiceSingle.mockResolvedValue(response);

    return store.dispatch(actions.getAmazonBook(book.isbn)).then(() => {
      expect(store.getActions()).toEqual(expectedResponse);
    });
  });
  it('can throw error on get amazon book', () => {
    const store = createMockStore();
    const response = null;
    const expectedResponse = [
      { isLoading: 'loading', type: 'FETCH_AMAZON_BOOK_IS_LOADING' },
      { isLoading: 'errored', type: 'FETCH_AMAZON_BOOK_IS_LOADING' },
      { hasErrored: true, type: 'FETCH_AMAZON_BOOK_HAS_ERRORED' },
    ];
    getAmazonBookServiceSingle.mockRejectedValue(response);

    return store.dispatch(actions.getAmazonBook(book.isbn)).then(() => {
      expect(store.getActions()).toEqual(expectedResponse);
    });
  });
});
