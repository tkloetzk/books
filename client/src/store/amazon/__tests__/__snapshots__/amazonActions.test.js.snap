// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`amazon actions can clear amazon books 1`] = `
Array [
  Array [
    Object {
      "type": "CLEAR_AMAZON_BOOKS_SUCCESS",
    },
  ],
]
`;
